### PR TITLE
JAMES-3371 JMAP Email/query Allow listing email

### PR DESCRIPTION
**Pull request by remk**

### CODE REVIEW

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+    server.getProbe(classOf[MailboxProbeImpl]).createMailbox(MailboxPath.inbox(BOB))
+    val otherMailboxPath = MailboxPath.forUser(BOB, "other")
+    server.getProbe(classOf[MailboxProbeImpl]).createMailbox(otherMailboxPath)
+    val messageId1: MessageId = server.getProbe(classOf[MailboxProbeImpl])
+      .appendMessage(BOB.asString, MailboxPath.inbox(BOB), AppendCommand.from(message))
+      .getMessageId
+
+    val messageId2: MessageId = server.getProbe(classOf[MailboxProbeImpl])
+      .appendMessage(BOB.asString, MailboxPath.inbox(BOB), AppendCommand.from(message))
+      .getMessageId
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala:79]: The 2 msgs are in the same mbx which do not match the test name...

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+        .header(ACCEPT.toString, ACCEPT_RFC8621_VERSION_HEADER)
+        .body(request)
+        .when
+        .post
+        .`then`
+        .statusCode(SC_OK)
+        .contentType(JSON)
+        .extract
+        .body
+        .asString
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala:166]: Indent of restassured

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+
+      val expectedResult = List(composedMessageId1, composedMessageId2, composedMessageId3)
+        .sortBy(_.getUid)
+        .map(_.getMessageId)
+        .map(_.serialize())
+
+      assertThatJson(Json.parse(response).\\("ids").head).isEqualTo(Json.toJson(expectedResult))
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala:236]: Use assertthatjson inPath to be more readable

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/EmailQuery.scala [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+import org.apache.james.mailbox.model.{MailboxId, MessageId}
+
+case class EmailQueryRequest(accountId: AccountId, inMailbox: Option[MailboxId])
+
+case class QueryState(value: String) extends AnyVal
+
+object QueryState {
+  def forIds(ids: Seq[MessageId]): QueryState = {
+
+    QueryState(
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/EmailQuery.scala:33]: Linebreak and {} are not needed?

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/EmailQuery.scala [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        .hashUnencodedChars(ids.map(_.serialize()).mkString(" "))
+        .toString)
+  }
+}
+
+case class EmailQueryResponse(accountId: AccountId,
+                              queryState: QueryState,
+                              canCalculateChanges: Boolean,
+                              ids: Seq[MessageId],
+                              position: Int)
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/EmailQuery.scala:44]: Strong type

Comment (chibenwa) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/EmailQuery.scala:44]: Return the limit actually enforced by the server (256) if applied, in order to be spec compliant?

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+         |    "c1"]]
+         |}""".stripMargin
+
+    awaitAtMostTenSeconds.untilAsserted { () =>
+      val response = `given`
+        .header(ACCEPT.toString, ACCEPT_RFC8621_VERSION_HEADER)
+        .body(request)
+        .when
+        .post.prettyPeek()
+        .`then`
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala:100]: indents when and then
debug spotted line above

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+        .sortBy(_.getUid)
+        .map(_.getMessageId)
+        .map(_.serialize())
+
+      assertThatJson(Json.parse(response).\\("ids").head).isEqualTo(Json.toJson(expectedResult))
+    }
+  }
+
+  @Test
+  def shouldListMailsShouldBeIdempotent(server: GuiceJamesServer): Unit = {
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala:241]:
```suggestion
  def listMailsShouldBeIdempotent(server: GuiceJamesServer): Unit = {
```

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala [hunk 8]
```diff
@@ -0,0 +1,304 @@
...
+        .contentType(JSON)
+        .extract
+        .body
+        .asString
+
+      assertThatJson(responseFirstCall).isEqualTo(responseSecondCall)
+    }
+  }
+
+  private def generateQueryState(messages: MessageId*): String = {
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala:299]: Why not reusing instead your def `forIds` declared in your QueryState companion object?

Comment (remk) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala:299]: I wanted to enforce the result of the hash in the contract. 
But it is open to discussion :)

Comment (chibenwa) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala:299]: I'm ok with @remk proposal

Comment (Arsnael) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala:299]: Maybe it's me not understanding something but... your code here:
```
    Hashing.murmur3_32()
      .hashUnencodedChars(messages.toList.map(_.serialize()).mkString(" "))
      .toString
```
and in QueryState:
```
 Hashing.murmur3_32()
        .hashUnencodedChars(ids.map(_.serialize()).mkString(" "))
        .toString)
```
Looks quite the same to me, except for the fact that here you use a vararg as param and a Seq in QueryState ? Aren't both sequential? 

I don't mind it too much at the end, just curious :)

Comment (remk) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala:299]: Yes it is exactly the same :)

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala [hunk 9]
```diff
@@ -0,0 +1,304 @@
...
+        .when
+        .post.prettyPeek()
+        .`then`
+        .statusCode(SC_OK)
+        .contentType(JSON)
+        .extract
+        .body
+        .asString
+
+      assertThatJson(response).withThreadDumpOnError().isEqualTo(
```
Comment (rouazana) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala:107]: is this `withThreadDumpOnError` really useful?

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailQueryMethod.scala [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+  override val requiredCapabilities: Capabilities = Capabilities(CORE_CAPABILITY, MAIL_CAPABILITY)
+
+  override def process(capabilities: Set[CapabilityIdentifier], invocation: Invocation, mailboxSession: MailboxSession, processingContext: ProcessingContext): Publisher[(Invocation, ProcessingContext)] =
+    metricFactory.decoratePublisherWithTimerMetricLogP99(JMAP_RFC8621_PREFIX + methodName.value,
+      asEmailQueryRequest(invocation.arguments)
+        .flatMap(processRequest(mailboxSession, invocation, _))
+        .onErrorResume({
+          case e: IllegalArgumentException => SMono.just(Invocation.error(ErrorCode.InvalidArguments, e.getMessage, invocation.methodCallId))
+          case e: Throwable => SMono.raiseError(e)
+        })
```
Comment (rouazana) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailQueryMethod.scala:51]: do you need the parenthesis around the accolades here?

**Comment by remk** [comment]
```
[7e7c31b75a6c8e231a524061370a10c323fb4163] [INFO] 
[7e7c31b75a6c8e231a524061370a10c323fb4163] [ERROR] Errors: 
[7e7c31b75a6c8e231a524061370a10c323fb4163] [ERROR]   RabbitMQTerminationSubscriberTest.lambda$eventProcessingShouldNotCrashOnInvalidMessage$1:112 » ConcurrentModification
[7e7c31b75a6c8e231a524061370a10c323fb4163] [INFO] 
[7e7c31b75a6c8e231a524061370a10c323fb4163] [ERROR] Tests run: 79, Failures: 0, Errors: 1, Skipped: 0
```

**Review by rouazana — APPROVED** [review]

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala [hunk 11]
```diff
@@ -0,0 +1,306 @@
...
+      .setSubject("test")
+      .setBody("testmail", StandardCharsets.UTF_8)
+      .build
+    server.getProbe(classOf[MailboxProbeImpl]).createMailbox(MailboxPath.inbox(BOB))
+    val otherMailboxPath = MailboxPath.forUser(BOB, "other")
+    server.getProbe(classOf[MailboxProbeImpl]).createMailbox(otherMailboxPath)
+    val messageId1: MessageId = server.getProbe(classOf[MailboxProbeImpl])
+      .appendMessage(BOB.asString, MailboxPath.inbox(BOB), AppendCommand.from(message))
+      .getMessageId
+    Thread.sleep(1000)
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala:76]: Why a sleep between the 2 mailboxes?

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala [hunk 12]
```diff
@@ -0,0 +1,306 @@
...
+import org.awaitility.Awaitility
+import org.awaitility.Duration.ONE_HUNDRED_MILLISECONDS
+import org.junit.jupiter.api.{BeforeEach, Test}
+import play.api.libs.json.Json
+
+trait EmailQueryMethodContract {
+
+  private lazy val slowPacedPollInterval = ONE_HUNDRED_MILLISECONDS
+  private lazy val calmlyAwait = Awaitility.`with`.pollInterval(slowPacedPollInterval).and.`with`.pollDelay(slowPacedPollInterval).await
+  private lazy val awaitAtMostTenSeconds = calmlyAwait.atMost(10, TimeUnit.SECONDS)
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala:48]: Just a question: why a lazy val?

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala [hunk 13]
```diff
@@ -0,0 +1,306 @@
...
+import org.apache.james.utils.DataProbeImpl
+import org.awaitility.Awaitility
+import org.awaitility.Duration.ONE_HUNDRED_MILLISECONDS
+import org.junit.jupiter.api.{BeforeEach, Test}
+import play.api.libs.json.Json
+
+trait EmailQueryMethodContract {
+
+  private lazy val slowPacedPollInterval = ONE_HUNDRED_MILLISECONDS
+  private lazy val calmlyAwait = Awaitility.`with`.pollInterval(slowPacedPollInterval).and.`with`.pollDelay(slowPacedPollInterval).await
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala:47]: IMO indent can be improved here

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala [hunk 14]
```diff
@@ -0,0 +1,306 @@
...
+      .setSubject("test")
+      .setBody("testmail", StandardCharsets.UTF_8)
+      .build
+    server.getProbe(classOf[MailboxProbeImpl]).createMailbox(MailboxPath.inbox(BOB))
+    val otherMailboxPath = MailboxPath.forUser(BOB, "other")
+    server.getProbe(classOf[MailboxProbeImpl]).createMailbox(otherMailboxPath)
+    val messageId1 = server.getProbe(classOf[MailboxProbeImpl])
+      .appendMessage(BOB.asString, MailboxPath.inbox(BOB), AppendCommand.from(message))
+      .getMessageId
+    Thread.sleep(1000)
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala:201]: Ah these sleeps are for ordering on receivedAt property. Maybe we can lower the value, and add a comment.

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailQueryMethod.scala [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ * *
+ * http://www.apache.org/licenses/LICENSE-2.0                 *
+ * *
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailQueryMethod.scala:11]: @Arsnael  will complain about license formatting

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala (hunk 11 above)
Comment (chibenwa) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala:76]: I found out ^^ https://github.com/linagora/james-project/pull/3770/files/2cad72754b338070956c34d85f8f41b574a66ea4#diff-d155e6ae7df11901205940a5d44d2a87

**Review by chibenwa — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala (hunk 12 above)
Comment (remk) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala:51]: To fix a java interop issue.
when defining them as 'val' i add an issue with java class implementing the contract. (not implemented abstract class)

using ' def'' or 'lazy val' fixed the issue

**Comment by remk** [comment]
could be considered green ?
```
[dd8a1a43996162a5b053c921257a90b2577294ff] [INFO] Apache James :: Server :: Web Admin server integration tests :: Memory SUCCESS [02:03 min]
[dd8a1a43996162a5b053c921257a90b2577294ff] [INFO] ------------------------------------------------------------------------
[dd8a1a43996162a5b053c921257a90b2577294ff] [INFO] BUILD SUCCESS
[dd8a1a43996162a5b053c921257a90b2577294ff] [INFO] ------------------------------------------------------------------------
[dd8a1a43996162a5b053c921257a90b2577294ff] [INFO] Total time:  02:38 h
[dd8a1a43996162a5b053c921257a90b2577294ff] [INFO] Finished at: 2020-09-11T06:19:51Z
```

failed deleting docker images

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**